### PR TITLE
backup: pipeline scan and save step

### LINKED ISSUE #11350: backup: run rocksDB operations (which are blocking) in separate worker pool
## Development Task
After #11158, the 'save file' part has been refactored to asynchronous functions. However, the `scan` part (which involves rocksDB operations, which includes blocking local disk I/O.) is still blocking and CPU bound. 
Some test cases shows that, with `noop://` backend, one backup worker thread can utilize 100% of CPU.

image

For now, we have a `semaphore` to limit the concurrency of backup tasks, but backup threads keep the `permit` when they are doing async i/o too:

Synchronous asynchronous I_O

We can reduce the problem by separating the control thread pool into two parts:

1. A tiny (maybe 1 or 2 threads) for running real async tasks (write, scan regions, etc.).
2. A worker pool (maybe `yatp::ThreadPool`) for scan tasks.

Task list:

- [ ] decouple scanning and writing.
- [ ] make local backend use `tokio::fs::File` instead of `std::fs::File`. (which is essential, or when we backing up to NFS, the blocking fs operations may block everything)
- [ ] add a new config for changing the tokio runtime size (not allowed to be modified in runtime).

### PR DESCRIPTION
**Pull request by YuJuncen**
### What is changed and how it works?

Close #11350.

What's Changed:

Added a new worker pool, and run I/O in it.

### TODO

- [x] ~~Make local file I/O (`local://` storages) async.~~ (This has already be done when refactoring backup to async...)
- [x] Add config terms.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- PR to update `pingcap/tidb-ansible`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test (The old test cases can help)
- Manual test (add detailed scripts or steps below)

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/backup/src/endpoint.rs [hunk 1]
```diff
@@ -151,6 +152,118 @@ pub struct BackupRange {
...
+
+impl KvWriter {
+    async fn save(self, storage: &dyn ExternalStorage) -> Result<Vec<File>> {
+        match self {
+            Self::TiDB(writer) => writer.save(storage).await,
+            Self::Raw(writer) => writer.save(storage).await,
+        }
+    }
+
+    fn need_flush_keys(&self) -> bool {
```
Comment (3pointer) [inline components/backup/src/endpoint.rs:170]: why not use this before send to `InMemBackupFiles` channel?

**Comment by YuJuncen** [comment]
```
[2021-12-15T05:44:27.172Z] thread 'main' panicked at 'assertion failed: `(left != right)`
[2021-12-15T05:44:27.172Z]   left: `0`,
[2021-12-15T05:44:27.172Z]  right: `0`', components/backup/src/endpoint.rs:1507:13
```
🤔

File: components/backup/src/endpoint.rs [hunk 2]
```diff
@@ -151,6 +152,118 @@ pub struct BackupRange {
...
+    start_key: Vec<u8>,
+    end_key: Vec<u8>,
+    start_version: TimeStamp,
+    end_version: TimeStamp,
+    region: Region,
+}
+
+impl std::fmt::Debug for InMemBackupFiles {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("InMemBackupFiles")
```
Comment (kennytm) [inline components/backup/src/endpoint.rs:189]: should just implement a dummy `Debug` for `KvWriter` and `#[derive(Debug)]` for `InMemBackupFiles`.

File: components/backup/src/endpoint.rs [hunk 3]
```diff
@@ -151,6 +152,118 @@ pub struct BackupRange {
     cf: CfName,
 }
 
+/// The generic saveable writer. for generic `InMemBackupFiles`.
+/// Maybe what we really need is make Writer a trait...
+enum KvWriter {
+    TiDB(BackupWriter),
```
Comment (kennytm) [inline components/backup/src/endpoint.rs:158]: this should be `Txn` not `TiDB`? the backup API isn't TiDB-specific.

File: components/backup/src/endpoint.rs [hunk 4]
```diff
@@ -293,33 +389,35 @@ impl BackupRange {
                 return Err(e);
             }
         }
+        drop(snap_store);
+        let stat = scanner.take_statistics();
+        if start_scan.saturating_elapsed_secs() > 30.0 {
```
Comment (kennytm) [inline components/backup/src/endpoint.rs:394]: place `start_scan.saturating_elapsed_secs()` into a variable. the elapsed duration is not a pure function.

**Comment by YuJuncen** [comment]
[2021-12-20T06:39:59.159Z] thread 'main' panicked at 'assertion failed: `(left != right)`
[2021-12-20T06:39:59.159Z]   left: `0`,
[2021-12-20T06:39:59.159Z]  right: `0`', components/backup/src/endpoint.rs:1513:13

Which means RocksDB may not trigger any disk I/O, can we just commit this?

**Comment by kennytm** [comment]
what do you mean by "commit this"

**Comment by YuJuncen** [comment]
> what do you mean by "commit this"

Just comment the assertion of `stats.fetch(IOType::Export, IOOp::Read) == 0`.

BTW, I guess this would be a bug introduced by the PR(Which made the statistic invalid.): during backup, the task may be scheduled to other threads (which doesn't set the thread local I/O type by `file_system::set_io_type`.).

**Comment by kennytm** [comment]
ok you meant "comm<b>ent</b> (out)" not "comm<b>it</b>".

File: components/backup/src/metrics.rs [hunk 5]
```diff
@@ -41,4 +41,21 @@ lazy_static! {
...
+    )
+    .unwrap();
+    pub static ref BACKUP_SCAN_KV_COUNT : IntCounterVec = register_int_counter_vec!(
+        "tikv_backup_scan_kv_count",
+        "Total number of kvs backed up",
+        &["cf"],
+    )
+    .unwrap();
+    pub static ref BACKUP_SCAN_KV_SIZE : IntCounterVec = register_int_counter_vec!(
+        "tikv_backup_scan_kv_size",
```
Comment (kennytm) [inline components/backup/src/metrics.rs:56]:
```suggestion
        "tikv_backup_scan_kv_size_bytes",
```

following the naming convention https://prometheus.io/docs/practices/naming/#metric-names

File: components/backup/src/endpoint.rs [hunk 6]
```diff
@@ -1401,7 +1501,7 @@ pub mod tests {
             assert!(!resp.has_error(), "{:?}", resp);
             let file_len = if *len <= SHORT_VALUE_MAX_LEN { 1 } else { 2 };
             let files = resp.get_files();
-            info!("{:?}", files);
+            println!("{:?}", files);
```
Comment (kennytm) [inline components/backup/src/endpoint.rs:1504]: revert?

**Review by kennytm — APPROVED** [review]
rest lgtm

**Review by 3pointer — APPROVED** [review]
LGTM

**Pull request merged**